Remove shop on app/uninstalled webhook (#16)
Shopify/turbolinks-jwt-sample-app: README.md

5. Your app is now able to access the authenticated `ProductsController` from the `HomeController` using session tokens.

### Marking shops as uninstalled on app/uninstalled webhook

> Uninstall webhook subscriptions and cleanup are a prerequisite for using session tokens. With session tokens, if the app does not clean up on install, the same shop cannot re-install the app successfully. The problem occurs because upon re-install, the shop record still exists on the app DB. So the app (via shopify_app gem) thinks the app is actually still installed on the shop and tries to skip OAuth and jump right into the app.

Reviewer: ```suggestion
To ensure OAuth still works with session tokens, your app must update its shop records in the event a shop uninstalls your app. You can be notified of these events by subscribing to the `app/uninstall` [webhook](https://shopify.dev/docs/admin-api/rest/reference/events/webhook).
```

5. Your app is now able to access the authenticated `ProductsController` from the `HomeController` using session tokens.

### Marking shops as uninstalled on app/uninstalled webhook

> To ensure OAuth still works with session tokens, your app must update its shop records in the event a shop uninstalls your app. You can be notified of these events by subscribing to the `app/uninstall` [webhook](https://shopify.dev/docs/admin-api/rest/reference/events/webhook).

Reviewer 2: ```suggestion
To ensure OAuth continues to work with session tokens, your app must update its shop records in the event a shop uninstalls your app. The app can receive notifications of these events by subscribing to the `app/uninstall` [webhook](https://shopify.dev/docs/admin-api/rest/reference/events/webhook).
```

Reviewer 2: Would it be worth giving an example of a scenario where OAuth and session tokens need to work together?

Maybe this README is not the right place for it. But I wonder if the connections/distinctions between OAuth, Session tokens/JWT, and cookies is clear. Something to consider explaining further in shopify.dev docs.

Author: Yeah that makes sense to add that in the `shopify.dev` guides.

#### Setup app/uninstall webhook

> 1. Use the `add_webhook` generator from ShopifyApp to setup the `app/uninstalled` webhook for the app.

Reviewer 2: Is this referring to the ShopifyApp cli or the gem? Can we add a link to the source appropriately, for convenience?

Author: This is referring to the gem. I just added a link to it

```shell script
rails g shopify_app:add_webhook -t app/uninstalled -a {your_app_url}/webhooks/app_uninstalled
```

> This will add the `app/uninstalled` webhook to your config and create the `AppUninstalledJob` job class for you.  

Reviewer 2: The suggestion below might not be capturing the correct detail. 
I am hoping we more specifically address "which config is getting updated", "where will the job be created", and "what will that let me do" for clarity.

```suggestion
This will add the `app/uninstalled` webhook to your app config and create the `AppUninstalledJob` job class for you to add uninstallation handler logic.  
```

2. Update the `AppUninstalledJob` job class to mark the Shop record as uninstalled.

```ruby
class AppUninstalledJob < ActiveJob::Base
  def perform(args)
    shop = Shop.find_by(shopify_domain: args[:shop_domain])

    mark_shop_as_uninstalled(shop)
  end

  private

  def mark_shop_as_uninstalled(shop)

>     shop.destroy! if shop

Reviewer 3: I wonder if we should make this demonstration of soft-deleting the shop so that any additional data in the shop model isn't lost.

Reviewer 2: +1. That would be very helpful to give app devs who currently rely on data persisting through reinstalls have a clear/reliable path forward.

Reviewer 4: We can't do soft delete. Shopify app doesn't use the soft delete gem and soft delete uses a different table structure.

Author: This would involve adding a variation from what shopify_app provides as well. I don't think the app should be deviating from default behaviour.

Reviewer 2: > Shopify app doesn't use the soft delete gem

There's a separate gem for soft deletion? :o

I think what Sam and I were referring to was simply "not permanently removing the shop record upon uninstall".
If this is not recommended, we should call it out in the docs as well since apps relying on persisting related data through reinstalls will be impacted by migrating to jwt.

Author: Agreed. I'm reaching out to DTE to see what their recommendation is for this.

  end
end
```

In this example, the app marks the shop as uninstalled by deleting the Shop record.

3. Define a background job to ensure existing shops also have the uninstall webhook setup.  

> In this app, the `RegisterWebhooksForActiveShop` job is defined to iterate all shop records in the database and configure the webhooks.

Reviewer 3: Should we add a note that this is only needed for apps that have already been installed on shops?

Reviewer 2: ```suggestion
3. Define a background job to ensure shops with existing installations of the app also have the uninstall webhook set up.  
In this app, the `RegisterWebhooksForActiveShop` job is defined to iterate all shop records in the database and configure the webhooks.
```

```ruby
class RegisterWebhooksForActiveShops < ApplicationJob
  queue_as :default

  def perform
    register_webhooks_for_active_shops
  end

  private

  def register_webhooks_for_active_shops
    Shop.find_each do |shop|
      ShopifyApp::WebhooksManagerJob.perform_now(
        shop_domain: shop.shopify_domain,
        shop_token: shop.shopify_token,

>         webhooks: ShopifyApp.configuration.webhooks

Reviewer 2: Are there any implications/considerations if `ShopifyApp.configuration.webhooks` contains webhooks other than uninstall? Specifically if any of the other topics in the array are already registered?

If yes, we should probably call that out and/or provide an alternative that targets only the new topics (for example maybe an inline array, assuming it is safe/acceptable practice)

```ruby
ShopifyApp::WebhooksManagerJob.perform_now(
  # ...
  webhooks: [
    {topic: 'app/uninstalled', address: "#{ENV['APP_URL']}/webhooks/app_uninstalled", format: 'json'},
  ]
)
```

Author: These webhook registrations [happen on every OAuth process](https://github.com/Shopify/shopify_app/blob/master/app/controllers/shopify_app/callback_controller.rb#L17). So my assumption is that it's idempotent. Going to confirm this with the webhooks team.
